File: CodingSamples/Fundamentals/Generic/EnumerableTest/DemoApp1/Common/SimpleStack.cs
namespace Common;

//a generic class with type parameter E
public class SimpleStack<E> : IStackReader<E>, IStackWriter<E>
{
    //nested class
    class Node
    {
        internal E element;
        internal Node below;

        internal Node Skip(int count)
        {
            Node n = this;
            for(int i = 0; i < count; ++i)
                n = n.below;
            return n;
        }
    }

    private Node top;

    public int Length { get; private set; }

    public void Push(E item)
    {
        top = new Node { element = item, below = top }; //instance initializer
        Length += 1;
    }

    public E Pop()
    {
        E item = top.element;
        top = top.below;
        Length -= 1;
        return item;
    }

    public Iterator GetEnumerator()
    {
        return new Iterator(this);
    }

    public struct Iterator(SimpleStack<E> source)
    {
        private Node next = source.top;

        public E Current { get; private set; }

        public bool MoveNext()
        {
            if(next != null)
            {
                Current = next.element;
                next = next.below;
                return true;
            }
            return false;
        }
    }

    //indexer - is a parameterized property for getting/setting 
    //a value in this instance at index specified by the parameter
    public E this[int index]
    {
        get{ return top.Skip(index).element; }
        set{ top.Skip(index).element = value; }
    }
}

File: CodingSamples/Java/Fundamentals/Generic/TypeParameterTest/demo2/common/SimpleStack.java
package common;

public class SimpleStack<V> {

    //inner member class
    class Node {

        V element;
        Node below;

        Node(V item){
            element = item;
            below = top;
        }
    }

    private Node top;

    public void push(V item) {
        top = new Node(item);
    }

    public V pop() {
        V item = top.element;
        top = top.below;
        return item;
    }

    public boolean empty() {
        return top == null;
    }

    public void copy(SimpleStack<? super V> target) {
        for(Node n = top; n != null; n = n.below)
            target.push(n.element);
    }

}

